File: web/src/lib/bridge/protocol.ts
export type WebDemoEventName =
  | "agent.turn.started"
  | "agent.text.delta"
  | "agent.reasoning.delta"
  | "agent.tool.call"
  | "agent.tool.result"
  | "agent.error"
  | "agent.turn.finish"
  | "permission.request"
  | "session.updated"
  | "session.expired"
  | "sandbox.diff.updated"
  | "sandbox.reset";

export type WebDemoEvent = {
  type: "event";
  event: WebDemoEventName;
  sessionId: string;
  turnId?: number;
  payload: Record<string, unknown>;
};

export type WebDemoRequestMethod =
  | "demo.session.create"
  | "turn.start"
  | "turn.interrupt"
  | "permission.resolve"
  | "sandbox.diff.get"
  | "sandbox.file.get"
  | "sandbox.reset"
  | "session.discard";

export type WebDemoRequest = {
  type: "request";
  id: string;
  method: WebDemoRequestMethod;
  sessionToken?: string;
  payload: Record<string, unknown>;
};

export type WebDemoResponse = {
  type: "response";
  id: string;
  ok: boolean;
  payload?: Record<string, unknown>;
  error?: {
    code: string;
    message: string;
  };
};

export type DemoSessionPayload = {
  sessionId: string;
  sessionToken: string;
  expiresAt: string;
  workspaceLabel: string;
  permissionMode: string;
  model: string;
};

export type TurnStartResult = {
  sessionId: string;
  turnId: number;
  finishReason: string;
};

export type ChangedFile = {
  path: string;
  status: string;
  additions: number;
  deletions: number;
};

export type SandboxDiffPayload = {
  changedFiles: ChangedFile[];
  patchAvailable: boolean;
  patch?: string;
  diskUsageBytes: number;
};

export type PermissionResolveBehavior = "allow" | "deny";
export type PermissionRememberScope = "" | "session" | "workspace";

export type PermissionResolvePayload = {
  requestId: string;
  behavior: PermissionResolveBehavior;
  rememberScope: PermissionRememberScope;
  message?: string;
};

export type WebDemoClientLike = {
  connect: () => Promise<void>;
  createSession: (accessCode: string) => Promise<DemoSessionPayload>;
  startTurn: (
    text: string,
    options: { sessionToken: string; permissionMode: string }
  ) => Promise<TurnStartResult>;
  resolvePermission: (
    sessionToken: string,
    payload: PermissionResolvePayload
  ) => Promise<boolean>;
  getSandboxFile: (sessionToken: string, path: string) => Promise<{ path: string; content: string }>;
  resetSandbox: (sessionToken: string) => Promise<boolean>;
  close: () => void;
};

export function isWebDemoEvent(value: unknown): value is WebDemoEvent {
  if (!value || typeof value !== "object") {
    return false;
  }
  const candidate = value as Partial<WebDemoEvent>;
  return candidate.type === "event" && typeof candidate.event === "string";
}

export function isWebDemoResponse(value: unknown): value is WebDemoResponse {
  if (!value || typeof value !== "object") {
    return false;
  }
  const candidate = value as Partial<WebDemoResponse>;
  return candidate.type === "response" && typeof candidate.id === "string";
}

export function readString(payload: Record<string, unknown>, key: string): string {
  const value = payload[key];
  return typeof value === "string" ? value : "";
}

export function readNumber(payload: Record<string, unknown>, key: string): number {
  const value = payload[key];
  return typeof value === "number" ? value : 0;
}

export function readRecord(payload: Record<string, unknown>, key: string): Record<string, unknown> {
  const value = payload[key];
  if (value && typeof value === "object" && !Array.isArray(value)) {
    return value as Record<string, unknown>;
  }
  return {};
}

export function readChangedFiles(payload: Record<string, unknown>): ChangedFile[] {
  const value = payload.changedFiles;
  if (!Array.isArray(value)) {
    return [];
  }
  return value.flatMap((item) => {
    if (!item || typeof item !== "object") {
      return [];
    }
    const record = item as Record<string, unknown>;
    const path = readString(record, "path");
    if (!path) {
      return [];
    }
    return [
      {
        path,
        status: readString(record, "status") || "modified",
        additions: readNumber(record, "additions"),
        deletions: readNumber(record, "deletions")
      }
    ];
  });
}

File: src/voice_code/web_demo/protocol.py
"""Typed protocol contracts for the web demo bridge."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

WebDemoEventName = Literal[
    "agent.turn.started",
    "agent.text.delta",
    "agent.reasoning.delta",
    "agent.tool.call",
    "agent.tool.result",
    "agent.error",
    "agent.turn.finish",
    "permission.request",
    "session.updated",
    "session.expired",
    "sandbox.diff.updated",
    "sandbox.reset",
]
WebDemoRequestMethod = Literal[
    "demo.session.create",
    "turn.start",
    "turn.interrupt",
    "permission.resolve",
    "sandbox.diff.get",
    "sandbox.file.get",
    "sandbox.reset",
    "session.discard",
]
PermissionResolveBehavior = Literal["allow", "deny"]
PermissionRememberScope = Literal["", "session", "workspace"]


class WebDemoEvent(BaseModel):
    """Stable event envelope consumed by the browser demo."""

    model_config = ConfigDict(populate_by_name=True)

    type: Literal["event"] = "event"
    event: WebDemoEventName
    session_id: str = Field(alias="sessionId")
    turn_id: int | None = Field(default=None, alias="turnId")
    payload: dict[str, object] = Field(default_factory=dict)


class WebDemoRequest(BaseModel):
    """Request sent by the web demo frontend over WebSocket."""

    model_config = ConfigDict(populate_by_name=True)

    type: Literal["request"] = "request"
    id: str
    method: WebDemoRequestMethod
    session_token: str = Field(default="", alias="sessionToken")
    payload: dict[str, object] = Field(default_factory=dict)


class WebDemoError(BaseModel):
    code: str
    message: str


class WebDemoResponse(BaseModel):
    """Stable response envelope for web demo requests."""

    type: Literal["response"] = "response"
    id: str = ""
    ok: bool
    payload: dict[str, object] | None = None
    error: WebDemoError | None = None


class PermissionResolvePayload(BaseModel):
    """Browser decision for a pending permission request."""

    model_config = ConfigDict(populate_by_name=True)

    request_id: str = Field(alias="requestId")
    behavior: PermissionResolveBehavior
    remember_scope: PermissionRememberScope = Field(default="", alias="rememberScope")
    message: str = ""


def error_response(code: str, message: str, *, request_id: str = "") -> WebDemoResponse:
    return WebDemoResponse(
        id=request_id,
        ok=False,
        error=WebDemoError(code=code, message=message),
    )
